Simplify IN subquery for sublink pullup.

Aegeaner opened the pull request:
In GP7's convert_ANY_sublink_to_join(), there is no simplification of origin subquery, as a result sometimes the sublink cannot be pulled up, which produced a bad query plan. In GP5, there is a subquery simplify step which helped sublink pull-up:
```
/* Delete ORDER BY and DISTINCT. */
		cdbsubselect_drop_orderby(subselect);
		cdbsubselect_drop_distinct(subselect);
```
As only simple sub-query is allowed to be pulled up:
```
                 /*
		 * Under certain conditions, we cannot pull up the subquery as a join.
		 */
		if (!is_simple_subquery(root, subselect, NULL, NULL))
			return NULL;
```

Take a look at the following query case:
```
EXPLAIN SELECT rmv_1.cover_no,
            rmv_1.period_no,
            rmv_1.risk_no,
            rmv_1.risk_period_no,
            rmv_1.location_ext_no,
            rmv_1.source_system
           FROM ctx.gis_risk_mv rmv_1
          WHERE ((rmv_1.cover_no, rmv_1.period_no, rmv_1.risk_no, rmv_1.risk_period_no, (rmv_1.source_system)::text) IN ( SELECT rp1.cover_no,
                    rp1.period_no,
                    rp1.risk_no,
                    rp1.risk_period_no,
                    rmv_1.source_system
                   FROM ( SELECT rp.cover_no,
                            rp.period_no,
                            rp.risk_no,
                            rp.risk_period_no,
                            row_number() OVER (PARTITION BY rp.cover_no, rp.period_no, rp.risk_no ORDER BY rp.risk_period_no DESC) AS risk_period_rank
                           FROM ctx.gis_risk_period rp) rp1
                  WHERE (rp1.risk_period_rank = 1)
                  GROUP BY rp1.cover_no, rp1.period_no, rp1.risk_no, rp1.risk_period_no, rmv_1.source_system));
```

But we lose this simplification since version 6X until master, which caused the sublink pull-up doesn't work:

```
Gather Motion 3:1  (slice5; segments: 3)  (cost=0.00..1334551428791.85 rows=77599 width=48)
   ->  Subquery Scan on rmv_1  (cost=0.00..1334551428791.85 rows=25867 width=48)
         Filter: (SubPlan 1)
         ->  Append  (cost=0.00..3478.96 rows=51733 width=210)
               ->  Seq Scan on sor_cm_risk_mv rskmv  (cost=0.00..236.89 rows=3097 width=208)
               ->  Seq Scan on sor_cm_risk_mv rskmv_1  (cost=0.00..225.56 rows=2719 width=208)
               ->  Seq Scan on sor_cm_risk_mv rskmv_2  (cost=0.00..1.00 rows=1 width=824)
               ->  Seq Scan on sor_cm_risk_mv rskmv_3  (cost=0.00..3015.51 rows=45917 width=210)
         SubPlan 1  (slice5; segments: 3)
           ->  HashAggregate  (cost=8599060.81..8599091.54 rows=1025 width=46)
                 Group Key: rp1.cover_no, rp1.period_no, rp1.risk_no, rp1.risk_period_no, rmv_1.source_system
                 ->  Subquery Scan on rp1  (cost=7446600.66..8598676.79 rows=10241 width=14)
                       Filter: (rp1.risk_period_rank = 1)
                       ->  WindowAgg  (cost=7446600.66..8214651.41 rows=10240677 width=14)
                             Partition By: rp.cover_no, rp.period_no, rp.risk_no
                             Order By: rp.risk_period_no
                             ->  Sort  (cost=7446600.66..7523405.74 rows=10240677 width=14)
                                   Sort Key: rp.cover_no, rp.period_no, rp.risk_no, rp.risk_period_no
                                   ->  Subquery Scan on rp  (cost=0.00..1080045.60 rows=10240677 width=14)
                                         ->  Append  (cost=0.00..772825.30 rows=10240677 width=319)
                                               ->  Result  (cost=0.00..170393.60 rows=1433280 width=319)
                                                     ->  Materialize  (cost=0.00..170393.60 rows=1433280 width=319)
                                                           ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..103754.40 rows=1433280 width=319)
                                                                 ->  Seq Scan on sor_cm_risk_period rskp  (cost=0.00..103754.40 rows=1433280 width=319)
                                               ->  Result  (cost=0.00..118234.50 rows=993900 width=319)
                                                     ->  Materialize  (cost=0.00..118234.50 rows=993900 width=319)
                                                           ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=0.00..72023.00 rows=993900 width=319)
                                                                 ->  Seq Scan on sor_cm_risk_period rskp_1  (cost=0.00..72023.00 rows=993900 width=319)
                                               ->  Result  (cost=0.00..94955.35 rows=777497 width=319)
                                                     ->  Materialize  (cost=0.00..94955.35 rows=777497 width=319)
                                                           ->  Broadcast Motion 3:3  (slice3; segments: 3)  (cost=0.00..58805.90 rows=777497 width=319)
                                                                 ->  Seq Scan on sor_cm_risk_period rskp_2  (cost=0.00..58805.90 rows=777497 width=319)
                                               ->  Result  (cost=0.00..865375.00 rows=7036000 width=319)
                                                     ->  Materialize  (cost=0.00..865375.00 rows=7036000 width=319)
                                                           ->  Broadcast Motion 3:3  (slice4; segments: 3)  (cost=0.00..538242.00 rows=7036000 width=319)
                                                                 ->  Seq Scan on sor_cm_risk_period rskp_3  (cost=0.00..538242.00 rows=7036000 width=319)
 Optimizer: Postgres query optimizer
(37 rows)
```
After we brought back `cdbsubselect_drop_distinct` and `cdbsubselect_drop_orderby`, we can got simplified subquery again and do the sublink pull-up optimization, which produced a better plan:
```
Gather Motion 3:1  (slice3; segments: 3)  (cost=8599137.66..8845829.72 rows=9700 width=48)
   ->  Hash Join  (cost=8599137.66..8845829.72 rows=3234 width=48)
         Hash Cond: ((rmv_1.cover_no = rp1.cover_no) AND (rmv_1.period_no = rp1.period_no) AND (rmv_1.risk_no = rp1.risk_no) AND (rmv_1.risk_period_no = rp1.risk_period_no))
         ->  Subquery Scan on rmv_1  (cost=0.00..6582.90 rows=51733 width=48)
               ->  Append  (cost=0.00..5030.93 rows=51733 width=210)
                     ->  Seq Scan on sor_cm_risk_mv rskmv  (cost=0.00..236.89 rows=3097 width=176)
                     ->  Seq Scan on sor_cm_risk_mv rskmv_1  (cost=0.00..225.56 rows=2719 width=176)
                     ->  Seq Scan on sor_cm_risk_mv rskmv_2  (cost=0.00..1.00 rows=1 width=792)
                     ->  Seq Scan on sor_cm_risk_mv rskmv_3  (cost=0.00..3015.51 rows=45917 width=178)
         ->  Hash  (cost=8599076.20..8599076.20 rows=1025 width=14)
               ->  Redistribute Motion 3:3  (slice2; segments: 3)  (cost=8598984.01..8599076.20 rows=1025 width=14)
                     Hash Key: rp1.cover_no, rp1.period_no, rp1.risk_no, rp1.risk_period_no
                     ->  HashAggregate  (cost=8598984.01..8599014.74 rows=1025 width=14)
                           Group Key: rp1.cover_no, rp1.period_no, rp1.risk_no, rp1.risk_period_no
                           ->  Subquery Scan on rp1  (cost=7446600.66..8598676.79 rows=10241 width=14)
                                 Filter: (rp1.risk_period_rank = 1)
                                 ->  WindowAgg  (cost=7446600.66..8214651.41 rows=10240677 width=14)
                                       Partition By: rp.cover_no, rp.period_no, rp.risk_no
                                       Order By: rp.risk_period_no
                                       ->  Sort  (cost=7446600.66..7523405.74 rows=10240677 width=14)
                                             Sort Key: rp.cover_no, rp.period_no, rp.risk_no, rp.risk_period_no
                                             ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..1694486.20 rows=10240677 width=14)
                                                   Hash Key: rp.cover_no, rp.period_no, rp.risk_no
                                                   ->  Subquery Scan on rp  (cost=0.00..1080045.60 rows=10240677 width=14)
                                                         ->  Append  (cost=0.00..772825.30 rows=10240677 width=319)
                                                               ->  Seq Scan on sor_cm_risk_period rskp  (cost=0.00..103754.40 rows=1433280 width=319)
                                                               ->  Seq Scan on sor_cm_risk_period rskp_1  (cost=0.00..72023.00 rows=993900 width=319)
                                                               ->  Seq Scan on sor_cm_risk_period rskp_2  (cost=0.00..58805.90 rows=777497 width=319)
                                                               ->  Seq Scan on sor_cm_risk_period rskp_3  (cost=0.00..538242.00 rows=7036000 width=319)
 Optimizer: Postgres query optimizer
(30 rows)
```

Notice this PR doesn't break the test cases in #12724 which fixed issue #12656, because we only do the simplification in the correlated case:
```
        if (correlated)
	{
		/* Delete ORDER BY and DISTINCT. */
		cdbsubselect_drop_orderby(subselect);
		cdbsubselect_drop_distinct(subselect);
```
Issue #12656 meets an explicitly distinct case, which will not go into this branch, so there would be no test case broken.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [x] Review a PR in return to support the community

kainwen commented:
PR pipelien fails due to cost mode change the added cases' agg shape, it is not a big problem. Please find some method to make it stable.

------------

Thanks a lot for your PR.

The code change LGTM.

Please:

1. states backgroup (the behavior of gp5) in the commit message
2. demonstrate what is the bad plan without your patch
3. add comments in the code change (why the change, why it is correct)

kainwen commented on src/backend/cdb/cdbsubselect.c:
```diff
@@ -464,7 +464,38 @@ SubqueryToJoinWalker(Node *node, ConvertSubqueryToJoinContext *context)
+/*
+ * cdbsubselect_drop_distinct
+ */
+void
```
I think we need some justification of the correctness and why doing so will help gpdb.

The comments in the below function is too straightforward.

Aegeaner replied:
I have added some comments inside the `convert_ANY_sublink_to_join()` function.

SmartKeyerror commented on src/test/regress/sql/subselect_gp.sql:
```diff
@@ -745,6 +745,23 @@ EXPLAIN SELECT '' AS five, f1 AS "Correlated Field"
   WHERE (f1, f2) IN (SELECT f2, CAST(f3 AS int4) FROM SUBSELECT_TBL
                      WHERE f3 IS NOT NULL) ORDER BY 2;
 
+EXPLAIN SELECT '' AS six, f1 AS "Correlated Field", f2 AS "Second Field"
```
The modification has changed the DISTINCT CLAUSE, but I did not see the corresponding test cases in this modification.

Aegeaner replied:
Thanks, but the test cases in line L1183~L1210 have already covered the distinct clause in subquery's case:
https://github.com/greenplum-db/gpdb/blob/000760776b6eac1d7da98eeacd56a89e9c7703cb/src/test/regress/sql/subselect_gp.sql#L1183~L1210

SmartKeyerror reviewed (commented):
When I apply this change to verify something, I got an error in the simple ANY correlated query:

```
create table t(a int, b int);
create table s(a int, b int);

insert into t values (1, 2), (2, 3), (3, 4);

insert into s values (1, 2), (2, 3), (3, 4);
insert into s values (1, 2), (2, 3), (3, 4);

explain analyze select * from t where a > (select a from s where s.a = t.a);
ERROR:  more than one row returned by a subquery used as an expression  (seg0 slice1 127.0.1.1:7002 pid=3155061)
```

SmartKeyerror commented:
@kainwen Sorry, I forget the data, the entire steps to reproduce is:

```
create table t(a int, b int);
create table s(a int, b int);

insert into t values (1, 2), (2, 3), (3, 4);

insert into s values (1, 2), (2, 3), (3, 4);
insert into s values (1, 2), (2, 3), (3, 4);

explain analyze select * from t where a > (select a from s where s.a = t.a);
```

SmartKeyerror commented:
Here is a recommendation, maybe we can use a more simple case to illustrate the pulled-up query tree and non-pulled-up query tree, the demo case in the merge messages is too complicated to read and check.

kainwen reviewed (approved):
LGTM.

Recommend to refine comments around the code.

Aegeaner commented:
> Here is a recommendation, maybe we can use a more simple case to illustrate the pulled-up query tree and non-pulled-up query tree, the demo case in the merge messages is too complicated to read and check.

Thanks for reviewing, I have simplified the test cases, removed unnecessary `ORDER BY 2,3` clauses at the top level of original queries, to make them more clear, and the query plans looks better now.

gpopt reviewed (approved):
+1

gpopt commented on src/backend/optimizer/plan/subselect.c:
```diff
@@ -1454,6 +1454,20 @@ convert_ANY_sublink_to_join(PlannerInfo *root, SubLink *sublink,
+		 *
+		 * There is no need to do the group-by or order-by inside the
+		 * subquery, if we have decided to pull up the sublink (when the
+		 * subquery is correlated). For the group-by case, after the sublink
```
Why does it have to be correlated?
We can drop them even the subquery is not correlated, right?

SmartKeyerror replied:
That's not right, we can't drop it if it's uncorrelated, more details can be found at #12724

gpopt replied:
AH, thanks, first time to see the usage of `DISTINCT ON (...)`.
